### DISCUSSION TITLE
Some decisions to be made

### DISCUSSION THREAD

**Discussion by BobanSpasic** [probonopd/MiniDexed#295] [General]
Hi @all

I would like to ask the developers about couple of points that needs to be decided at some point in time (better sooner than later).
- usage of Yamaha SysEx ID and the ID for dumps
- performance files

I apologize in advance of being a smart-ass (not being a developer in this project, but coming with such topics anyway), but I think I should share some experience with the others here.

**Yamaha SysEx ID**
At the moment, MiniDexed is responding to Yamaha SysEx messages. The pros are that you can use any editor for Yamaha DX7 to work with MiniDexed. The cons - you are bound to the existing format and can't change anything about it.
Yamaha's own solution for maintaining the backward compatibility are additions to the base SysEx dumps where each Synth model reads just the part he recognizes (where DX7 reads just the base format and ignores the rest, DX7II reads the base and supplement formats etc., and MiniDexed is expecting just the base format without additional sections)
These additional parts have their own header identifications.
Why I am starting this topic?
- we should think about how to transfer MiniDexed performance files. If we remain on using the Yamaha SysEx ID, we need at least our own second-level header ID as an extension to f=$7E (Universal SysEx dump). For remaining Yamaha-compatible, it should be uppercase, 8 characters long ID (Yamaha is using IDs like LM__8973PE or LM__8023AE where the underscores are spaces, we could take something like LM__MDEXVO and LM__MDEXPE).
Anyway, dunno how Yamaha would react on messing with their property.

**Performance files**
The current MiniDexed Performance files have nothing in common with DX-series Performance files. Nobody says they need to.
DX-series performance files are not storing the voice parameters. There are just the "links" to the voices with addition of performance parameters. If one edits the voice - the performance will also be affected.
MiniDexed Performance files are sort of Multi-Layer voice file. My vote goes for multi-layer instead of linked performance files.
Why I am starting this topic?
- we need a decision about the implementation of MIDI CC and SysEx for MiniDexed's performance files - the list of Parameters with CC implementation and a draft for SysEx dumps.

I think that these topics need to be decided soon, before the development of performance filed goes further in undefined direction.

**Comment by now-its-dark** [comment]
>`Nobody says they need to.`

FWIW, I do believe there is a meaningful workflow benefit to adhering to established spec whenever possible, as it ensures that the range of existing patches and software tools will be fully functional. 

There are some very well conceived and mature editors/librarians made specifically for 8-part multitimbral DX series synths, which utilize the "legacy" performance parameters. Even if some of the extended params that are unique to MiniDexed aren't present in those tools, it is so useful to be able to see nearly everything going on from one panel.

**Comment by probonopd** [comment]
@BobanSpasic you are correct that we want to be compatible with the original DX/TX series as much as pragmatically feasible. For example, it is a must that existing DX7 voice editors on the PC can work with MiniDexed. When it comes to the features beyond the DX7 (e.g., storing performances), MiniDexed is currently not following any TX model. This was mainly done for very pragmatic reasons (read: it was most convenient for the persons who have implemented the respective features to implement them in this way). I also happen to think that this is less critical since most editors I know can't edit performances, only voices, anyhow.

> There are some very well conceived and mature editors/librarians made specifically for 8-part multitimbral DX series synths, which utilize the "legacy" performance parameters. 

@now-its-dark Interesting. Do you have some links handy? Are those open source?

Doing it like real TX models would require additional work, and would require that you always need to have the matching voices (in the matching banks?) around if you want to use a performance. Would this a) be a worthwhile tradeoff and b) who would want to do the work?

**Reply by BobanSpasic** [reply]
DX/TX performances are a bit limited and aren't really applicable to MiniDexed (microtuning and such). I would go for own performances format (poly/mono, portamento, midi channel etc per voice/ per TG, not per performance).
The editor is my part of the job, but before that - I would need a draft of MIDI implementation of the performance parameters (CC) and the format of an SysEx dump. 
Converting TX performances to MiniDexed performances would also be doable at some point.

**Reply by now-its-dark** [reply]
> Interesting. Do you have some links handy? Are those open source?

@probonopd There are a few applications which stood out to me: 

[Synlib DX-TX](http://m.bareille.free.fr/synlibdxtx/synlibdxtx.htm) is free, with source code available for the Atari ST version. There is a pretty nice windows version which runs well in WINE.

[Patch Base](https://coffeeshopped.com/patch-base#:~:text=TX7-,TX802,-TX816) is a paid app, but worth mentioning because it is a good option for iPad, in keeping with the "mini" aspect / portability factor and it supports numerous other synths as well. 

[Edisyn](https://github.com/eclab/edisyn) might be a good option for augmentation to support performances, since the DX7 single-voice part is in place already and there are multi-part editors for TX81Z and other multitimbral 4-OP FM, which could possibly serve as a template? 

My personal favorite multi-synth editor is SoundDiver, which is in the grey area of abandonware at this point... easy to find on archive.org if interested. it has a pretty deep editor / librarian adaptation for the TX802 and runs fine on modern OSs, natively on windows and stable in WINE. It's not pretty, but quite capable. (pictured below)
![SD_TX802](https://user-images.githubusercontent.com/5728860/176332434-799001a2-207f-49bc-b996-30c6e6732498.png)

**Comment by BobanSpasic** [comment]
I've taken a look at DX/TX SysEx messages for Performances just to find out that every model uses another implementation and the implementations are not compatible with each other. 
Please define an implementation for MiniDexed. I will be glad to help.

**Reply by probonopd** [reply]
Intuition tells me that it would be best to use TX802 sysex format for performances, but someone would have to implement it in MiniDexed...
